lib/streamlit/commands/query_params.py

```diff
@@ -140,3 +141,52 @@ def _ensure_no_embed_params(
         separator = "&" if current_embed_params else ""
         return separator.join([query_string, current_embed_params])
     return current_embed_params
+
+
+@dataclass
+class QueryParams:
+    """
+    A dict-like representation of query params that generally behaves like st.session.state.
+    The main difference is that it only stores and returns str and list[str].
+
+    TODO(willhuang1997): Fill in these docs with examples and fix doc above. Above is just a stub for now.
+    """
+
+    query_params: dict[str, List[Any]] = field(default_factory=dict)
+
+    def __init__(self, query_params: Dict[str, List[Any]] = {}):
+        self.query_params = query_params
+
+    def __repr__(self):
+        return util.repr_(self)
+
+    def __contains__(self, key: str) -> bool:
+        return key in self.query_params
+
+    def __len__(self):
+        return len(self.query_params)
+
+    def _check_ctx(self):
+        ctx = get_script_run_ctx()
+        if ctx is None:
+            return
+
+    def _send_query_params(self):
+        ctx = get_script_run_ctx()
+        msg = ForwardMsg()
+        msg.page_info_changed.query_string = _ensure_no_embed_params(
+            self.query_params, ctx.query_string
+        )
+        ctx.query_string = msg.page_info_changed.query_string
+        ctx.enqueue(msg)
+
+    def clear(self):
+        self._check_ctx()
```

Comment:
these methods call `ctx = get_script_run_ctx()` twice but my thoughts are that `_send_query_params` and `_check_ctx` will be used by `__setitem__` and `__setattr__` and so I think it makes it cleaner(?). However, happy to do something else if someone thinks otherwise as it is technically inefficient.